agent/config/file_watcher.go

```diff
@@ -0,0 +1,221 @@
+			w.reconcile()
+		case <-ctx.Done():
+			close(w.done)
+			return
+		}
+	}
+
+}
+
+func (w *FileWatcher) handleEvent(event fsnotify.Event) error {
+	w.logger.Debug("event received ", "filename", event.Name, "OP", event.Op)
+	// we only want Create and Remove events to avoid triggering a relaod on file modification
+	if !isCreate(event) && !isRemove(event) && !isWrite(event) && !isRename(event) {
+		return nil
+	}
+	filename := filepath.Clean(event.Name)
+	configFile, basename, ok := w.isWatched(filename)
+	if !ok {
+		return fmt.Errorf("file %s is not watched", event.Name)
+	}
+
+	// we only want to update inode and re-add if the event is on the watched file itself
+	if filename == basename {
+		if isRemove(event) {
+			// If the file was removed, try to reconcile and see if anything changed.
+			w.logger.Debug("attempt a reconcile ", "filename", event.Name, "OP", event.Op)
+			configFile.id = 0
+			w.reconcile()
+		}
+	}
+	if isCreate(event) || isWrite(event) || isRename(event) {
+		w.logger.Debug("call the handler", "filename", event.Name, "OP", event.Op)
+		go w.handleFunc(&WatcherEvent{Filename: filename})
+	}
+	return nil
+}
+
+func (w *FileWatcher) isWatched(filename string) (*watchedFile, string, bool) {
+	path := filename
+	configFile, ok := w.configFiles[path]
+	if ok {
+		return configFile, path, true
+	}
+
+	stat, err := os.Stat(filename)
+	if err != nil {
+		return nil, path, false
+	}
+	if !stat.IsDir() {
+		w.logger.Trace("not a dir")
+		// try to see if the watched path is the parent dir
+		NewPath := filepath.Dir(path)
+		w.logger.Trace("get dir", "dir", NewPath)
+		configFile, ok = w.configFiles[NewPath]
+	}
+	return configFile, path, ok
+}
+
+func (w *FileWatcher) reconcile() {
+	for filename, configFile := range w.configFiles {
+		w.logger.Debug("reconciling", "filename", filename)
+		newId, err := w.getFileId(filename)
+		if err != nil {
+			w.logger.Error("failed to get file id", "file", filename, "err", err)
+			continue
+		}
+
+		err = w.watcher.Add(filename)
+		if err != nil {
+			w.logger.Error("failed to add file to watcher", "file", filename, "err", err)
+			continue
+		}
+		if configFile.id != newId {
+			w.logger.Debug("call the handler", "filename", filename, "old id", configFile.id, "new id", newId)
+			w.configFiles[filename].id = newId
+			go w.handleFunc(&WatcherEvent{Filename: filename})
+		}
+	}
+}
+
+func isCreate(event fsnotify.Event) bool {
+	return event.Op&fsnotify.Create == fsnotify.Create
+}
+
+func isRemove(event fsnotify.Event) bool {
+	return event.Op&fsnotify.Remove == fsnotify.Remove
+}
+
+func isWrite(event fsnotify.Event) bool {
+	return event.Op&fsnotify.Write == fsnotify.Write
+}
+
+func isRename(event fsnotify.Event) bool {
+	return event.Op&fsnotify.Rename == fsnotify.Rename
+}
+
+func (w *FileWatcher) getFileId(filename string) (uint64, error) {
+	fileInfo, err := os.Stat(filename)
+	if err != nil {
+		return 0, err
+	}
+
+	return uint64(fileInfo.ModTime().Nanosecond()), err
```

Comment:
`(time.Time).Nanosecond` only returns the nanoseconds _within the current second_, not the total elapsed nanoseconds since 1970, that is `(time.Time).UnixNano`.

But a better question is why even convert this to a number at all? This function (were it renamed) could just return `(time.Time, error)` and then do comparisons using `(time.Time).{Equal, After, Before}` as necessary instead.